### PR TITLE
[NumPy]Set numpy default dtype

### PR DESCRIPTION
**Pull request by JiangZhaoh**
## Description ##
Add new feature to be consistent with numpy default type. 
For example, in original, default type only support 32 bit storage dtype such as float32 (or int32 in some special op). However, with this new feature, user can set 64 bit storage type as default dtype, such as float64 (int64 for some special op).
e.g.

```
import mxnet
from mxnet import numpy as np
from mxnet import npx
a=np.array([1, 2, 3])  # a.dtype = 'float32'
prev_state=npx.set_np_default_dtype()  # prev_state = False
b=np.array([1, 2, 3])  # b.dtype = 'float64'
```

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/numpy/random.py [hunk 1]
```diff
@@ -99,7 +99,8 @@ def uniform(low=0.0, high=1.0, size=None, dtype=None, ctx=None, out=None):
         ``low`` and ``high`` are both scalars. Otherwise,
         ``np.broadcast(low, high).size`` samples are drawn.
     dtype : {'float16', 'float32', 'float64'}, optional
-        Data type of output samples. Default is 'float32'
+        Data type of output samples. Default is 'float32' or 'float64', which
+        denends on your current default dtype.
```
Comment (haojin2) [inline python/mxnet/numpy/random.py]: probably say something like "when np_default_type=True" to clearly state which setting will result in which data type.

File: src/operator/numpy/np_init_op.cc [hunk 2]
```diff
@@ -57,6 +57,20 @@ inline bool NumpyIndicesShape(const nnvm::NodeAttrs& attrs,
...
+inline bool NumpyIndicesType(const nnvm::NodeAttrs& attrs,
+                             std::vector<int>* in_attrs,
+                             std::vector<int>* out_attrs) {
+  const IndicesOpParam& param = nnvm::get<IndicesOpParam>(attrs.parsed);
+  CHECK_EQ(in_attrs->size(), 0U);
+  CHECK_EQ(out_attrs->size(), 1U);
+  TYPE_ASSIGN_CHECK(*out_attrs, 0, param.dtype == -1 ?
+                                   (Imperative::Get()->is_np_default_dtype() ?
+                                    mshadow::kInt64 :
+                                    mshadow::kInt32) :
```
Comment (haojin2) [inline src/operator/numpy/np_init_op.cc]: Directly use `int64`, no need for such logic

File: src/operator/tensor/init_op.h [hunk 3]
```diff
@@ -44,6 +44,19 @@
 namespace mxnet {
 namespace op {
 
+inline int GetDefaultDtype() {
```
Comment (haojin2) [inline src/operator/tensor/init_op.h]: better use `DType`

Comment (haojin2) [reply src/operator/tensor/init_op.h]: & maybe you can move this to `src/common/utils.h`

File: src/operator/tensor/init_op.cc [hunk 4]
```diff
@@ -92,6 +91,16 @@ NNVM_REGISTER_OP(_full)
   .set_attr<FCompute>("FCompute<cpu>", InitFillWithScalarCompute<cpu>)
 .add_arguments(InitOpWithScalarParam::__FIELDS__());
 
+NNVM_REGISTER_OP(_npi_full)
```
Comment (haojin2) [inline src/operator/tensor/init_op.cc]: please register this op in `src/operator/numpy/np_init_op.cc/cu`

File: python/mxnet/ndarray/numpy/_op.py [hunk 5]
```diff
@@ -600,9 +613,13 @@ def identity(n, dtype=None, ctx=None):
     if n < 0:
         raise ValueError("Input 'n' cannot be negative")
     if ctx is None:
-        ctx = current_context()
-    dtype = _np.float32 if dtype is None else dtype
-    return _npi.identity(shape=(n, n), ctx=ctx, dtype=dtype)
+        ctx = str(current_context())
+    else:
+        ctx = str(ctx)
```
Comment (haojin2) [inline python/mxnet/ndarray/numpy/_op.py:618]: is `str` required here?

Comment (JiangZhaoh) [reply python/mxnet/ndarray/numpy/_op.py:618]: Refer to other similiar works, I think `str` is required here.

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [windows-cpu]

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [windows-cpu]

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [windows-cpu, windows-gpu]

**Comment by yzhliu** [comment]
@leezu @szha do you have example to enable naive engine in ci?

**Comment by leezu** [comment]
You can export the respective environment variable in the CI environment. Later @szha will enable it by default.

But we shouldn't use CI to debug the issue. To debug this issue I suggest to run the tests locally and do a binary over the set of enabled numpy tests, minimizing the number of tests run to trigger the bug.

File: src/operator/tensor/init_op.cc [hunk 6]
```diff
@@ -150,6 +148,16 @@ NNVM_REGISTER_OP(_linspace)
 .set_attr<FCompute>("FCompute<cpu>", LinspaceCompute<cpu>)
 .add_arguments(RangeParam::__FIELDS__());
 
+NNVM_REGISTER_OP(_npi_linspace)
```
Comment (haojin2) [inline src/operator/tensor/init_op.cc:151]: also move this to `src/operator/numpy/np_init_op.cc`

File: src/operator/tensor/init_op.cu [hunk 7]
```diff
@@ -70,6 +70,9 @@ NNVM_REGISTER_OP(_contrib_arange_like)
 NNVM_REGISTER_OP(_linspace)
 .set_attr<FCompute>("FCompute<gpu>", LinspaceCompute<gpu>);
 
+NNVM_REGISTER_OP(_npi_linspace)
```
Comment (haojin2) [inline src/operator/tensor/init_op.cu:73]: move this to `src/operator/numpy/np_init_op.cu`

File: src/operator/tensor/init_op.h [hunk 8]
```diff
@@ -288,12 +289,25 @@ struct InitOpWithScalarParam : dmlc::Parameter<InitOpWithScalarParam> {
...
-    DMLC_DECLARE_FIELD(dtype).set_default(mshadow::kFloat32)
+    DMLC_DECLARE_FIELD(dtype)
+      .set_default(-1)
+      .add_enum("None", -1)
       MXNET_ADD_ALL_TYPES_WITH_BOOL
       .describe("Target data type.");
     DMLC_DECLARE_FIELD(value)
       .describe("Value with which to fill newly created tensor");
   }
+  void SetAttrDict(std::unordered_map<std::string, std::string>* dict) {
```
Comment (haojin2) [inline src/operator/tensor/init_op.h:300]: one more blank line above.

**Pull request merged**